bpo-45321: Add missing error codes to module `xml.parsers.expat.errors`

hartwork opened the pull request:
Hi CPython!

This is my first pull request to CPython, so please bear with me, I'm happy to learn and adjust this pull request as needed.

To fix bug https://bugs.python.org/issue45321, this pull request gets the 6 error codes missing from module `xml.parsers.expat.errors` both added and documented.

Description "This should not be reported to Python applications" (at two places) is inspired by existing error code `XML_ERROR_ABORTED` (see https://github.com/python/cpython/blame/f76889a88720b56c8174f26a20a8e676a460c7a6/Doc/library/pyexpat.rst#L857).

I'm looking forward to your review and comments.

Best, Sebastian

PS: I am involved with [libexpat upstream](https://github.com/libexpat/libexpat), in case that matters for anything.

<!-- issue-number: [bpo-45321](https://bugs.python.org/issue45321) -->
https://bugs.python.org/issue45321
<!-- /issue-number -->

hartwork committed:
pyexpat: Add missing error codes to xml.parsers.expat.errors (bpo-45321)

https://bugs.python.org/issue45321

arhadthedev commented on Doc/library/pyexpat.rst:
```diff
@@ -867,6 +867,40 @@ The ``errors`` module has the following attributes:
+
+.. data:: XML_ERROR_NO_BUFFER
+
+   This should not be reported to Python applications.
```
`XML_ERROR_NO_BUFFER` → [`MemoryError`](https://docs.python.org/3/library/exceptions.html#MemoryError)

hartwork replied:
There seems to be a misunderstanding of `XML_ERROR_NO_BUFFER` here. it is not an out-of-memory situation but a situation where there is no buffer at all. And it won't be thrown in CPython because of how the current integration calls libexpat.

arhadthedev commented on Doc/library/pyexpat.rst:
```diff
@@ -867,6 +867,40 @@ The ``errors`` module has the following attributes:
+
+.. data:: XML_ERROR_INVALID_ARGUMENT
+
+   This should not be reported to Python applications.
```
`XML_ERROR_INVALID_ARGUMENT` → [`ValueError`](https://docs.python.org/3/library/exceptions.html#ValueError)

hartwork replied:
I don't see how `XML_ERROR_INVALID_ARGUMENT` would be thrown in CPython because of how the current integration calls libexpat.

arhadthedev reviewed (commented):
Some exceptions marked as "This should not be reported to Python applications" already have Python counterparts so they can be removed from documentation and ADD_CONST. Instead, they should be caught and rethrown:

corona10 commented on Modules/pyexpat.c:
```diff
@@ -1654,7 +1654,18 @@ static int
+
+    /* Be robust towards runtime libexpat being too old to know
+     * all the latest error codes */
+    const char *error_string = error_string_or_null;
```
@hartwork 
Out of curiosity, since CPython vendors the `libexpat` with a specific version.
Is there a possibility of adding unexpected constants that XML_ErrorString will return null?

hartwork replied:
Hi @corona10 , very valid question! `XML_ErrorString` returns `NULL` whenever the implementation does not know the error code that is being queried. That happens (a) for invalid values and (b) for values from the future, e.g. when calling `XML_ErrorString(XML_ERROR_AMPLIFICATION_LIMIT_BREACH)` with libexpat <2.4.0.

In general, there is (1) the libexpat version used at _build time_ and (2) the expat version used at _runtime_, loaded through dynamic linking. And they can be different. So in my mind there are two questions:

- Is there a chance that someone builds CPython with (vulnerable) libexpat <2.4.0 (and is it okay to fail the build if so)?
- Is there a chance that someone uses CPython binaries with libexpat binaries other than (and older than) the ones that CPython was built against (and is it okay to fail loading module `xml.parsers.expat.errors` if so)?

For status quo of this pull request I decided I would want to err on the side of caution for the case with older runtime. For compile time, it can be argued that libexpat <2.4.0 is vulnerable, and that hence the person compiling against libexpat <2.4.0 needs to get off of their vulnerable copy anyway. But that's my personable take, I'm happy to discuss. We could protect against both or none of these two scenarios. What do you think?

corona10 replied:
> Is there a chance that someone builds CPython with (vulnerable) libexpat <2.4.0 (and is it okay to fail the build if so)?

My personal opinion is okay to fail because the current implementation does not protect such situations,
if such uses cases were a lot, related issues are already created but it looks not.

But let's hear @vstinner and @tiran 's opinion for both questions, they know more about a lot of related history.

Merry Christmas.

hartwork replied:
> the current implementation does not protect such situations

@corona10 I think I should note that that the current implementation only uses error codes that are so old, it doesn't really have that problem. But adding new codes changes the game and raises these two new questions to be answered one way or another.

Merry Christmas to you and your family.

corona10 replied:
@hartwork 
> I think I should note that that the current implementation only uses error codes that are so old,

Ah I see, in that case, what about adding version checking macro for those constants, or what about detecting such constants by macro and adding constants at compile time?
I didn't check libexpat can use such approaches.

see similar approaches: https://github.com/python/cpython/blob/078abb676cf759b1e960f78390b6e80f256f0255/Modules/_gdbmmodule.c#L128

https://github.com/python/cpython/blob/078abb676cf759b1e960f78390b6e80f256f0255/configure.ac#L5888-L5900

hartwork replied:
@corona10 building pre-processor guards from https://github.com/libexpat/libexpat/blob/9542ad247130801eb226225299d01a17c0582f08/expat/lib/expat.h#L1042-L1044 would be easy, but it would mean that the set of available codes now become "dynamic" in the sense that the user can no longer assume that all those error code constants are available in module `xml.parsers.expat.errors`. Maybe that's worse than making the build require libexpat >=2.4.0 (some clean way). What do you think?

corona10 replied:
@hartwork 
> Maybe that's worse than making the build require libexpat >=2.4.0 (some clean way). 

If you worry about custom libexpat build, IMHO dynamic constant adding is useful and we can announce it the documentation. 

But I concur with the fixed version requirement, but it will need some discussion with other core devs.
You can use a mailing list or Discourse (https://discuss.python.org/c/core-dev/).
Since you are libexpat module maintainer, I think that your opinion will be persuasive.

corona10 commented on Modules/pyexpat.c:
```diff
@@ -1654,7 +1654,18 @@ static int
+     * all the latest error codes */
+    const char *error_string = error_string_or_null;
+    char error_string_substitute[100];
+    if (error_string_or_null == NULL) {
```
We can use the previous code right since version guard prevents the situation?

`````c
const char *error_string = XML_ErrorString(value);
`````

hartwork replied:
This is where the difference between runtime and build time comes into play, as explained at https://github.com/python/cpython/pull/30188#discussion_r775037301 . I'm good either way, I just need a decision if you want robust or simple.

corona10 replied:
I prefer the following style but I am not an expert on distro's issues.
But due to the distro's issue with build and runtime inconsistency problem, if we need a fallback string, I am okay with the change.
So I would like to request the code review for this part to @tiran and @vstinner :)

`````c
const char *error_string = XML_ErrorString(value);
if (error_string == NULL) {
    return -1;
}
if (PyModule_AddStringConstant(errors_module, name, error_string) < 0) {
    return -1;
}
`````

hartwork replied:
> I prefer the following style [..]

This will make the module import fail. That seems suboptimal. What's the idea, why the preference?
For proof:
```console
# ./python -c 'import xml.parsers.expat.errors'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "[..]/cpython/Lib/xml/parsers/expat.py", line 4, in <module>
    from pyexpat import *
    ^^^^^^^^^^^^^^^^^^^^^
SystemError: execution of module pyexpat failed without setting an exception
```

corona10 replied:
@hartwork 
Import will fail if build environment != runtime environment. right?
It will fail if the environment is different, I just want to know the CPython of the distro version can be executed with the different runtime version.
If the normal distro version guarantees the built environment and runtime environment are the same until the user does not tweak libexpat version, it will be okay.
IMHO(It's a really personal idea), if users want to run a different libexpat version, they should rebuild CPython and it looks like the right way.

Early fail which I suggested is a pretty usual workaround in CPython module initialization.

`````

@@ -1658,12 +1658,9 @@ add_error(PyObject *errors_module, PyObject *codes_dict,

     /* Be robust towards runtime libexpat being too old to know
      * all the latest error codes */
-    const char *error_string = error_string_or_null;
-    char error_string_substitute[100];
-    if (error_string_or_null == NULL) {
-        sprintf(error_string_substitute,
-                "error code %i too recent for runtime libexpat", value);
-        error_string = error_string_substitute;
+    const char *error_string = XML_ErrorString(value);
+    if (error_string == NULL) {
+        return -1;
     }

     if (PyModule_AddStringConstant(errors_module, name, error_string) < 0) {

$> cpython git:(pr/30188) ✗ ./python.exe -c 'import xml.parsers.expat.errors'
`````

corona10 replied:
So if distro version is built with vendor version, and run with system installed library version,
my suggestion will be wrong.

hartwork replied:
@corona10 I think the most likely case is building CPython with `--with-system-expat` and then using the resulting CPython binaries with older libexpat binaries at runtime, e.g. in context of a binary Linux distro with multiple release branches. If everyone agrees that that scenario is meant to be 100% unsupported and okay break to pyexpat, then we can just go with `return -1` on a `NULL` error string.

corona10 replied:
@hartwork 
I check the Linux distro which uses libexpat 1.6.7 it works as expected (module import success + use errr code for specific version)

I would like to suggest to changing it to return -1.

hartwork replied:
> @hartwork I check the Linux distro which uses libexpat 1.6.7 it works as expected (module import success + use errr code for specific version)

Thanks for giving it a spin.
1.6.7 is a is a shared library file version, not a version of libexpat, and not trivial to "decode". Let's stick to release and packaging versions please, so that everyone knows which versions are being referred to.


> I would like to suggest to changing it to return -1.

Please elaborate why. Thanks.

corona10 replied:
@hartwork 

My opinion is that unsupporting error codes should not be added.
https://github.com/python/cpython/blob/196b53eb1e62871ca80dee180e4891b4dd5c52ac/Modules/_sqlite/module.c#L303
For example, sqlite3 uses a similar approach, the only difference is they do not extract value at runtime.
If you think that the error_string should be filled even if the runtime environment is different. Please go ahead.
But does it guarantee normal behavior? if so I am okay with filling fallback string. 
Even if we choose any options I will leave this PR to other core dev to listen to their opinion also.
Since this patch will affect a lot of distros.

hartwork replied:
Hi @corona10 ,

I have dropped the `NULL` handling for now, because there are multiple places where `XML_ErrorString` is called and if someone decided that a runtime-build-time-mismatch is a problem, all of these places would need `NULL` handling:

```console
# git --no-pager grep 'XML_ErrorString(' -- Modules/pyexpat.c
Modules/pyexpat.c:                                  XML_ErrorString(code), lineno, column);
Modules/pyexpat.c:    return Py_BuildValue("z", XML_ErrorString((int)code));
Modules/pyexpat.c:    const char *error_string = XML_ErrorString(value);
```

> https://github.com/python/cpython/blob/196b53eb1e62871ca80dee180e4891b4dd5c52ac/Modules/_sqlite/module.c#L303
> 
> For example, sqlite3 uses a similar approach, the only difference is they do not extract value at runtime.

That runtime part difference is key, though. With only compile time, there cannot be a build/run mismatch.

> If you think that the error_string should be filled even if the runtime environment is different. Please go ahead.
> But does it guarantee normal behavior? if so I am okay with filling fallback string.

Interesting point, I guess that's a question of view.

corona10 reviewed (approved):
LGTM, but waiting for at least one more core dev's approvement.

scoder reviewed (changes requested):
There is no precedence in optionally available error constants here and I do not see a reason to start doing that. I think we should make them always available, regardless of the expat version.

hartwork committed:
pyexpat: Make xml.parsers.expat.errors robust towards outdated libexpat (bpo-45321)

The idea is to ensure that module `xml.parsers.expat.errors`
contains all known most error codes and messages,
even when CPython is compiled or run with an outdated versions
of libexpat.

This work is preparing for the upcoming addition of missing
error codes from libexpat >=2.0.0 and even >=2.4.0.

https://bugs.python.org/issue45321

scoder commented on Modules/pyexpat.c:
```diff
@@ -1650,16 +1650,93 @@ add_submodule(PyObject *mod, const char *fullname)
+     * our copy of error message */
+    const char * const runtime_error_string = XML_ErrorString((int)error_code);
+    assert((runtime_error_string == NULL) || ! strcmp(runtime_error_string, error_string));
+#endif
```
So, if you feel like asserting this (in debug mode), why not just use the runtime provided string, if available?

The error strings are owned by libexpat, not by CPython. IMHO, it's ok to provide a fallback so that we can always provide such a string, even if the runtime libexpat does not know it – although that seems somewhat useless since that error will then also not occur. But we should otherwise rely on libexpat to (potentially) know things better than we do, and ask it.

hartwork replied:
Hi @scoder,

what would be the advantage of making `XML_ErrorString` the runtime source of truth?
Since pyexpat passes a _static_ set of integers, is there any gain in information?

If there is any concern about the strings changing in libexpat in the future (of which I have no plans), we should not go store a copy in pyexpat. If we do store a copy, pyexpat becomes the new source of truth.  Everyone running libexpat 2.2.x in the RHEL 8 with six of these error strings only available in pyexpat would probably agree that pyexpat has become the new source of truth.

Maybe the debugging code block was a bit out of place anyway. I have dropped it now, and would like to present the result as my current recommendation.

Best, Sebastian

hartwork replied:
PS: @scoder I realize now that our difference in view above is rather minor. I have implemented your preference now, or so I think. If you change your mind, I have backup branches to restore to, easily :beers:

scoder replied:
> If there is any concern about the strings changing in libexpat in the future (of which I have no plans), we should not go store a copy in pyexpat.

Well, `libexpat` _is_ the source of truth. Whatever we do in `pyexpat` should follow (unless there is a compelling reason not to).

That being said, I admit that the original design is a bit surprising here. What you would normally expect is an error code as a key to an error message, whether that message is part of the originally reported error "thing" (read, exception) itself or not. Here, the error _message_ is the key to the error code. Obviously, that is not something we can change at this point.

Now, even if there are no plans to change error messages, I'm sure there's precedence for such changes in other projects, so I wouldn't rule it out entirely. And it's not impossible that some users persisted errors in one way or another, so changing messages might break someone's data. But – I think that is not a decision for `pyexpat` to take but by `libexpat`, so:

> If we do store a copy, pyexpat becomes the new source of truth. Everyone running libexpat 2.2.x in the RHEL 8 with six of these error strings only available in pyexpat would probably agree that pyexpat has become the new source of truth.

I don't think so. Old versions of libexpat do not change any more, and we know their behaviour. We can happily provide fallback constants for them (especially since they won't trigger them). `pyexpat` should not become a source of truth at all, but it can become a source of backported functionality. That allows `pyexpat` users to write simple code that responds to specific errors even if they won't occur in the current runtime `libexpat` version.

scoder replied:
> PS: @scoder I realize now that our difference in view above is rather minor. I have implemented your preference now, or so I think. If you change your mind, I have backup branches to restore to, easily beers

Thanks. Yes, I think that's how it should work.

hartwork replied:
@scoder I think we have different meanings for "source of truth" in mind. Beyond that, I don't think we disagree. I'm optimistic that my latest changes in code already do what you wanted. Let me know.

corona10 reviewed (approved):
@scoder 
This looks better.
LGTM

erlend-aasland commented on Modules/pyexpat.c:
```diff
@@ -1699,53 +1778,16 @@ add_errors_module(PyObject *mod)
-        }                                                           \
-    } while(0)
+    size_t error_index = 0;
+    for (; error_index < sizeof(error_info_of) / sizeof(struct ErrorInfo); error_index++) {
```
FTR, there's a Py macro for this: `Py_ARRAY_LENGTH` ;)
